File: mdt-cli/mcp_client.py
import subprocess
import json
from typing import Dict, Any, Optional
from dataclasses import dataclass
from config import Config


@dataclass
class TicketData:
    """Data structure for ticket information."""
    key: str
    title: str
    type: str
    priority: str
    status: str
    description: Optional[str] = None
    rationale: Optional[str] = None


class MCPClientError(Exception):
    """Custom exception for MCP client errors."""
    pass


class MCPClient:
    """Handles communication with MCP server."""

    def __init__(self, server_path: str):
        self.server_path = server_path

    def get_ticket(self, project: str, key: str) -> Optional[TicketData]:
        """
        Retrieve complete ticket information from MCP server.

        Args:
            project: Project code (e.g., "MDT")
            key: Full ticket key (e.g., "MDT-066")

        Returns:
            TicketData object or None if not found
        """
        try:
            # Get basic ticket attributes
            attrs_text = self._call_mcp_tool("get_cr", {
                "project": project,
                "key": key
            })

            if not attrs_text:
                return None

            # Parse the formatted text response
            attrs = self._parse_cr_response(attrs_text, key)

            # Get sections for description and rationale
            sections = self._get_ticket_sections(project, key)

            return TicketData(
                key=attrs.get("key", key),
                title=attrs.get("title", "No Title"),
                type=attrs.get("type", "Unknown"),
                priority=attrs.get("priority", "Medium"),
                status=attrs.get("status", "Unknown"),
                description=sections.get("description"),
                rationale=sections.get("rationale")
            )

        except Exception as e:
            raise MCPClientError(f"Error fetching ticket {key}: {e}")

    def _get_ticket_sections(self, project: str, key: str) -> Dict[str, str]:
        """Extract description and rationale sections."""
        sections = {}

        try:
            # Look for description and rationale sections
            section_names = ["## Description", "## Rationale"]

            for section_name in section_names:
                try:
                    content = self._call_mcp_tool("get_cr_section", {
                        "project": project,
                        "key": key,
                        "section": section_name
                    })
                    if content:
                        # Extract actual content from formatted response
                        extracted_content = self._extract_section_content(content)
                        if extracted_content:
                            # Map to simple names
                            simple_name = section_name.replace("## ", "").lower()
                            sections[simple_name] = extracted_content
                except Exception as e:
                    # Continue with other sections if one fails
                    print(f"Warning: Could not fetch section {section_name} for {key}: {e}")

        except Exception as e:
            # Don't raise exception for section errors, just log warning
            print(f"Warning: Could not fetch sections for {key}: {e}")

        return sections

    def _extract_section_content(self, response_text: str) -> str:
        """Extract actual content from the formatted section response."""
        lines = response_text.split('\n')
        content_lines = []
        capture = False

        for line in lines:
            if line.strip() == '---':
                if capture:
                    # End of content
                    break
                else:
                    # Start of content
                    capture = True
                    continue
            elif capture:
                # Capture content lines
                if line.strip() and not line.startswith('📖') and not line.startswith('**Section:') and not line.startswith('**Content Length:'):
                    content_lines.append(line)

        return '\n'.join(content_lines).strip()

    def _parse_cr_response(self, response_text: str, key: str) -> Dict[str, str]:
        """Parse the formatted text response from get_cr tool."""
        attrs = {
            "key": key,
            "title": "No Title",
            "type": "Unknown",
            "priority": "Medium",
            "status": "Unknown"
        }

        lines = response_text.split('\n')
        current_field = None

        for line in lines:
            line = line.strip()

            # Extract title from the header line
            if line.startswith('📄 **') and ' - ' in line:
                # Format: 📄 **MDT-066** - Title
                parts = line.split(' - ', 1)
                if len(parts) == 2:
                    title = parts[1].rstrip('*')
                    attrs["title"] = title.strip()
                continue

            # Look for metadata fields
            if line.startswith('- Status:'):
                attrs["status"] = line.replace('- Status:', '').strip()
            elif line.startswith('- Type:'):
                attrs["type"] = line.replace('- Type:', '').strip()
            elif line.startswith('- Priority:'):
                attrs["priority"] = line.replace('- Priority:', '').strip()

        return attrs

    def _call_mcp_tool(self, tool_name: str, params: Dict[str, Any]) -> Any:
        """Execute MCP tool call using JSON-RPC protocol."""
        process = None
        try:
            # Create JSON-RPC request
            request = {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "tools/call",
                "params": {
                    "name": tool_name,
                    "arguments": params
                }
            }

            # Start MCP server process
            process = subprocess.Popen(
                ["node", self.server_path],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True
            )

            # Send JSON-RPC request
            request_json = json.dumps(request) + "\n"
            stdout, stderr = process.communicate(input=request_json, timeout=Config.get_timeout())

            if process.returncode != 0:
                error_msg = stderr.strip() if stderr else "Unknown error"
                raise MCPClientError(f"MCP server error: {error_msg}")

            # Parse JSON-RPC response
            try:
                # Handle multi-line output (server logs + JSON)
                lines = stdout.strip().split('\n')
                json_line = None
                for line in reversed(lines):  # Look from the end for JSON
                    if line.strip().startswith('{'):
                        json_line = line.strip()
                        break

                if not json_line:
                    raise MCPClientError("No JSON response found in MCP server output")

                response = json.loads(json_line)
                if "error" in response:
                    raise MCPClientError(f"MCP tool error: {response['error']}")
                if "result" in response:
                    result = response["result"]
                    # Handle MCP response format with content array
                    if isinstance(result, dict) and "content" in result:
                        if isinstance(result["content"], list) and len(result["content"]) > 0:
                            # Extract text from content items
                            text_content = ""
                            for item in result["content"]:
                                if isinstance(item, dict) and "text" in item:
                                    text_content += item["text"]
                            return text_content
                    return result
                else:
                    raise MCPClientError("Invalid MCP response format")
            except json.JSONDecodeError as e:
                raise MCPClientError(f"Invalid JSON response from MCP server: {e}")

        except subprocess.TimeoutExpired:
            if process:
                process.kill()
                process.wait()
            raise MCPClientError("MCP server timeout")
        except FileNotFoundError:
            raise MCPClientError(f"MCP server not found at {self.server_path}")
        except Exception as e:
            if process:
                process.kill()
                process.wait()
            raise MCPClientError(f"Unexpected error calling MCP server: {e}")
        finally:
            # Ensure process is always cleaned up
            if process and process.poll() is None:
                process.kill()
                process.wait()

File: mcp-server/tests/e2e/helpers/mcp-client.ts
/**
 * Real MCP Client Wrapper
 *
 * Provides a unified interface for communicating with MCP server
 * through both stdio and HTTP transports using proper MCP protocol.
 */

import type { ChildProcess } from 'node:child_process'
import type { TestEnvironment } from './test-environment'
import { MCPLogger } from './mcp-logger'
import { HttpTransport, StdioTransport } from './mcp-transports'

export interface MCPTool {
  name: string
  description?: string
  inputSchema?: Record<string, unknown>
}

export interface MCPResponse {
  success: boolean
  data?: unknown
  key?: string // For CR creation responses
  error?: {
    code: number
    message: string
  }
}

export interface MCPClientOptions {
  transport?: 'stdio' | 'http'
  timeout?: number
  retries?: number
}

export class MCPClient {
  private testEnv: TestEnvironment
  private options: MCPClientOptions
  private transport?: StdioTransport | HttpTransport
  private logger = MCPLogger.getInstance()
  private requestId = 0
  private serverProcess?: ChildProcess
  private pendingRequests: Map<number, {
    resolve: (value: Record<string, unknown>) => void
    reject: (reason: Error) => void
  }> = new Map()

  constructor(testEnv: TestEnvironment, options: MCPClientOptions = {}) {
    this.testEnv = testEnv
    this.options = {
      transport: options.transport || 'stdio',
      timeout: options.timeout || 10000,
      retries: options.retries || 3,
    }
    this.logger.info('MCPClient created', { transport: this.options.transport })
  }

  /**
   * Start the MCP client and server connection
   */
  async start(): Promise<void> {
    this.logger.info('Starting MCP client', { transport: this.options.transport })

    if (this.transport) {
      const error = new Error('Client already started')
      this.logger.error('MCP client already started', error)
      throw error
    }

    try {
      if (this.options.transport === 'stdio') {
        // Use MCP SDK client for stdio transport
        await this.startStdioClient()
      }
      else {
        // Use HTTP transport (existing implementation)
        this.transport = new HttpTransport(this.testEnv)
        await this.transport.start()
      }
      this.logger.info('MCP client started successfully', { transport: this.options.transport })
    }
    catch (error) {
      this.transport = undefined
      this.logger.error('Failed to start MCP client', error instanceof Error ? error : new Error(String(error)))
      throw new Error(`Failed to start MCP client: ${error instanceof Error ? error.message : String(error)}`)
    }
  }

  /**
   * Start stdio client with proper MCP protocol handling
   */
  private async startStdioClient(): Promise<void> {
    // Use the existing StdioTransport which spawns the server correctly
    this.transport = new StdioTransport(this.testEnv)
    await this.transport.start()

    // Get reference to the spawned process for direct communication
    this.serverProcess = (this.transport as StdioTransport).getProcess()

    // Set up proper stdio communication that handles MCP protocol
    this.setupStdioCommunication()
  }

  /**
   * Set up stdio communication to handle MCP protocol properly
   */
  private setupStdioCommunication(): void {
    if (!this.serverProcess) {
      throw new Error('Server process not available')
    }

    // The MCP server using StdioServerTransport expects JSON-RPC messages
    // Each message should be sent as a separate line followed by \n
    // We need to ensure proper message framing

    // Buffer for incoming data
    let buffer = ''

    this.serverProcess.stdout?.on('data', (data) => {
      buffer += data.toString()

      // Process complete lines
      const lines = buffer.split('\n')
      buffer = lines.pop() || '' // Keep incomplete line in buffer

      // Process each complete JSON-RPC message
      lines.forEach((line) => {
        if (line.trim()) {
          try {
            const message = JSON.parse(line)
            // Handle the message appropriately
            this.handleServerMessage(message)
          }
          catch {
            this.logger.warn('Failed to parse server message', { line: line.trim() })
          }
        }
      })
    })

    // Handle transport-level process errors once and fail all pending requests.
    this.serverProcess.on('error', (error) => {
      const pending = Array.from(this.pendingRequests.values())
      this.pendingRequests.clear()
      pending.forEach(({ reject }) => reject(error))
    })
  }

  /**
   * Handle incoming messages from the MCP server
   */
  private handleServerMessage(message: Record<string, unknown>): void {
    // Handle JSON-RPC response
    if (message.id && this.pendingRequests.has(message.id as number)) {
      const { resolve, reject } = this.pendingRequests.get(message.id as number)!
      this.pendingRequests.delete(message.id as number)

      if (message.error) {
        // Create a custom error that preserves the error code
        const errorObj = message.error as { message?: string, code?: number }
        const error = new Error(errorObj.message || 'RPC error') as Error & { code?: number }
        error.code = errorObj.code
        reject(error)
      }
      else {
        resolve(message as Record<string, unknown>)
      }
    }
  }

  /**
   * Stop the MCP client and server connection
   */
  async stop(): Promise<void> {
    this.logger.info('Stopping MCP client')

    if (this.transport) {
      try {
        await this.transport.stop()
        this.logger.info('MCP client stopped successfully')
      }
      catch (error) {
        this.logger.error('Error stopping transport', error instanceof Error ? error : new Error(String(error)))
      }
      finally {
        this.transport = undefined
        this.serverProcess = undefined
      }
    }
  }

  /**
   * Check if client is connected to server
   */
  async isConnected(): Promise<boolean> {
    return this.transport?.isConnected() || false
  }

  /**
   * List available tools from MCP server
   */
  async listTools(): Promise<MCPTool[]> {
    const response = await this.callTool('tools/list', {})
    if (!response.success || !response.data) {
      return []
    }

    // Handle different response formats
    if (Array.isArray(response.data)) {
      return response.data
    }

    // Type guard for object with tools property
    if (typeof response.data === 'object' && response.data !== null && 'tools' in response.data) {
      const dataWithTools = response.data as { tools?: unknown }
      if (Array.isArray(dataWithTools.tools)) {
        return dataWithTools.tools
      }
    }

    return []
  }

  /**
   * Call a specific MCP tool with retry logic
   */
  async callTool(toolName: string, params: Record<string, unknown>): Promise<MCPResponse> {
    const maxRetries = this.options.retries || 3
    const retryDelay = 1000 // 1 second between retries

    this.logger.debug('Calling MCP tool', { toolName, params, maxRetries })

    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        if (!this.transport || !this.transport.isConnected()) {
          const response = {
            success: false,
            error: {
              code: -1,
              message: 'Client not connected to server',
            },
          }
          this.logger.warn('Client not connected to server', { attempt, maxRetries })
          return response
        }

        // Prepare JSON-RPC request with unique ID using timestamp + counter
        const uniqueId = Date.now() * 1000 + (++this.requestId) // Ensure uniqueness even in concurrent calls
        const request = {
          jsonrpc: '2.0',
          id: uniqueId,
          method: toolName === 'tools/list' ? 'tools/list' : 'tools/call',
          params: toolName === 'tools/list'
            ? {}
            : {
                name: toolName,
                arguments: params,
              },
        }

        // Send request and get response based on transport
        let response: Record<string, unknown>
        if (this.options.transport === 'stdio') {
          response = await this.callToolStdio(request)
        }
        else {
          response = await this.callToolHttp(request)
        }

        // Parse response
        if (response.error) {
          const errorObj = response.error as { code?: number, message?: string }
          const mcpResponse = {
            success: false,
            error: {
              code: errorObj.code || -32000,
              message: errorObj.message || 'RPC error',
            },
          }

          // Special handling for rate limit errors - throw them for test compatibility
          if (errorObj.message
            && (errorObj.message.toLowerCase().includes('rate limit')
              || errorObj.message.includes('Rate limit'))) {
            this.logger.warn('Rate limit error encountered', {
              toolName,
              error: mcpResponse.error,
            })
            // Create a proper Error object for the test
            const rateLimitError = new Error(errorObj.message) as Error & { code?: number }
            rateLimitError.code = errorObj.code
            throw rateLimitError
          }

          // Check if error is retryable
          if (!this.isRetryableError(mcpResponse.error.code, mcpResponse.error.message) || attempt === maxRetries) {
            this.logger.warn('Tool call failed with non-retryable error', {
              toolName,
              error: mcpResponse.error,
            })
            return mcpResponse
          }

          // Wait before retrying
          this.logger.debug('Retrying tool call', {
            toolName,
            attempt,
            delay: retryDelay * attempt,
          })
          await this.sleep(retryDelay * attempt)
          continue
        }

        // Parse the result content first
        const content = this.parseResultContent(
          typeof response.result === 'object' && response.result !== null
            ? (response.result as Record<string, unknown>)
            : null,
        )

        // Check if the response indicates an error according to MCP spec
        if (response.result && (response.result as { isError?: boolean }).isError === true) {
          // According to MCP spec, this is a tool execution error
          // The error message should be in the content array
          let errorMessage = 'Tool execution error'
          if ((response.result as { content?: unknown[] }).content && Array.isArray((response.result as { content?: unknown[] }).content)) {
            const textContent = (response.result as { content: unknown[] }).content.find((c: unknown) => typeof c === 'object' && c !== null && (c as { type?: string }).type === 'text')
            if (textContent && (textContent as { text?: string }).text) {
              errorMessage = (textContent as { text: string }).text
            }
          }

          const mcpResponse = {
            success: false,
            error: {
              code: -32000, // Server error for tool execution errors
              message: errorMessage,
            },
          }

          this.logger.warn('Tool returned error with isError=true', {
            toolName,
            error: mcpResponse.error,
          })
          return mcpResponse
        }

        // Check if the content contains an error message in the expected format
        // Some tools (like get_cr, update_cr_status) return error messages as formatted content
        // with success=true, so we should NOT convert these to failures
        // Only check for errors if content is a string
        if (content && typeof content === 'string' && content.includes('❌ **Error in')) {
          // Check if this is a tool that returns errors as content
          const toolsWithContentErrors = ['get_cr', 'update_cr_status', 'create_cr', 'delete_cr', 'update_cr_attrs', 'manage_cr_sections', 'list_crs']

          // Extract the tool name from the error message
          const toolNameMatch = content.match(/❌ \*\*Error in (\w+)\*\*/)
          const errorToolName = toolNameMatch ? toolNameMatch[1] : ''

          // If this is a tool that returns errors as content, treat it as success
          if (toolsWithContentErrors.includes(errorToolName)) {
            const mcpResponse = {
              success: true,
              data: content,
            }
            this.logger.info('Tool returned error as formatted content', { toolName, errorToolName })
            return mcpResponse
          }

          // For other tools, convert to failure
          const errorMatch = content.match(/❌ \*\*Error in \w+\*\*[\s\S]+?\n\n(.+)(?:\n\n|$)/)
          const errorMessage = errorMatch ? errorMatch[1] : content

          const mcpResponse = {
            success: false,
            error: {
              code: -32000,
              message: errorMessage.trim(),
            },
          }

          this.logger.warn('Tool call failed with error in content', {
            toolName,
            error: mcpResponse.error,
          })
          return mcpResponse
        }

        // Success - return the content
        const mcpResponse = {
          success: true,
          data: content,
        }

        this.logger.info('Tool call successful', { toolName, attempt })
        return mcpResponse
      }
      catch (error: unknown) {
        this.logger.warn('Tool call threw exception', {
          toolName,
          attempt,
          error: error instanceof Error ? error.message : String(error),
        })

        // If this is the last attempt, return the error
        if (attempt === maxRetries) {
          // Preserve the error code from the server
          // Protocol errors should maintain their original codes (-32601, -32602, etc.)
          // Tool execution errors will have code -32000 by default
          const errorCode = (error instanceof Error && 'code' in error)
            ? ((error as Error & { code?: number }).code ?? -1)
            : -1

          const response: MCPResponse = {
            success: false,
            error: {
              code: errorCode,
              message: error instanceof Error ? error.message : 'Unknown error',
            },
          }
          this.logger.error('Tool call failed after all retries with exception', error instanceof Error ? error : new Error(String(error)), { toolName, maxRetries, errorCode },
          )
          return response
        }

        // Wait before retrying
        await this.sleep(retryDelay * attempt)
      }
    }

    // This should never be reached, but TypeScript requires it
    const response = {
      success: false,
      error: {
        code: -1,
        message: 'Max retries exceeded',
      },
    }
    this.logger.error('Unexpected error in callTool', new Error('Should not reach here'), { toolName, maxRetries })
    return response
  }

  /**
   * Call tool via stdio transport with proper MCP protocol
   */
  private async callToolStdio(request: Record<string, unknown>): Promise<Record<string, unknown>> {
    if (!this.serverProcess) {
      throw new Error('Server process not available')
    }

    // Ensure request.id is a number
    const requestId = typeof request.id === 'number' ? request.id : Date.now()
    const methodName = typeof request.method === 'string' ? request.method : 'unknown'

    return new Promise<Record<string, unknown>>((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(requestId)
        reject(new Error(`Tool call timeout: ${methodName}`))
      }, this.options.timeout)

      // Store the promise callbacks for response handling
      this.pendingRequests.set(requestId, {
        resolve: (response: unknown) => {
          clearTimeout(timeout)
          // Ensure response is Record<string, unknown>
          if (typeof response === 'object' && response !== null) {
            resolve(response as Record<string, unknown>)
          }
          else {
            reject(new Error('Invalid response type from server'))
          }
        },
        reject: (error: unknown) => {
          clearTimeout(timeout)
          if (error instanceof Error) {
            reject(error)
          }
          else {
            reject(new Error(String(error)))
          }
        },
      })

      // Send the request as a JSON-RPC message
      const message = `${JSON.stringify(request)}\n`
      this.serverProcess?.stdin?.write(message)
    })
  }

  /**
   * Call tool via HTTP transport
   */
  private async callToolHttp(request: Record<string, unknown>): Promise<Record<string, unknown>> {
    const baseUrl = (this.transport as HttpTransport).getBaseUrl()

    const response = await fetch(`${baseUrl}/mcp`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    })

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`)
    }

    return (await response.json()) as Record<string, unknown>
  }

  /**
   * Parse result content from MCP response
   */
  private parseResultContent(result: Record<string, unknown> | null): unknown {
    if (!result)
      return null

    // Handle MCP content array format: {content: [{text: "...", type: "text"}]}
    if (result.content && Array.isArray(result.content)) {
      const textItems = result.content
        .filter((item: unknown) => typeof item === 'object' && item !== null && (item as { type?: string }).type === 'text')
        .map((item: unknown) => (item as { text?: string }).text)

      const text = textItems.length === 1 ? textItems[0] : textItems.join('\n')

      return text
    }

    // Handle direct content
    const content = result.content || result

    // If content is not a string, convert it appropriately
    if (typeof content !== 'string' && typeof content !== 'undefined') {
      // For tools/list, we expect an array of tools
      if (Array.isArray(content)) {
        return content
      }
      // For other non-string content, return as-is (could be parsed JSON already)
      return content
    }

    // Return content as-is without automatic JSON parsing
    return content
  }

  /**
   * Check if an error code is retryable
   */
  private isRetryableError(errorCode?: number, errorMessage?: string): boolean {
    if (!errorCode)
      return false

    // Never retry intentional throttling responses.
    if (errorMessage && /rate limit/i.test(errorMessage)) {
      return false
    }

    // Retry only transport/connectivity failures.
    return errorCode === -1
  }

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms))
  }

  /**
   * Register a project for testing
   * This is a no-op method for test compatibility
   */
  registerProject(projectCode: string, projectInfo: { name: string, path: string, description?: string }): void {
    // In the actual MCP server implementation, projects are discovered
    // through the configuration system, not registered at runtime
    // This method exists for test compatibility only
    this.logger.debug('Project registration requested (no-op)', { projectCode, projectInfo })
  }
}
